File: controller/src/test/java/com/desktoppet/core/MountedBehaviorEngineTest.java
package com.desktoppet.core;

import com.desktoppet.model.Envelope;
import com.desktoppet.model.VoicePackAction;
import com.desktoppet.model.VoicePackGroup;
import com.desktoppet.model.VoicePackInfo;
import com.desktoppet.network.Protocol;
import com.google.gson.JsonObject;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

class MountedBehaviorEngineTest {

    private MountedBehaviorEngine engine;
    private Path basePath;

    @BeforeEach
    void setUp() {
        basePath = Path.of("C:", "fake", "voice-pack");

        List<VoicePackAction> tapHeadActions = List.of(
            new VoicePackAction(1, "motions/33.motion3.json", "audios/33.ogg", null, "Hello!", 1000L, 1000L),
            new VoicePackAction(2, "motions/34.motion3.json", null, null, "Hi!", 500L, 500L)
        );

        List<VoicePackAction> audioOnlyActions = List.of(
            new VoicePackAction(3, null, "audios/1.ogg", null, "Audio only", 1000L, 1000L)
        );

        Map<String, VoicePackGroup> groups = new LinkedHashMap<>();
        groups.put("tap_head", new VoicePackGroup("tap_head", "触摸头部", 2, tapHeadActions));
        groups.put("audio_only", new VoicePackGroup("audio_only", "Audio only", 2, audioOnlyActions));

        VoicePackInfo voicePack = new VoicePackInfo("test-pack", "Test Pack", "test-code", basePath, groups, List.of());
        engine = new MountedBehaviorEngine(voicePack);
    }

    @Test
    void hasGroupForArea_returnsTrueForKnownArea() {
        assertTrue(engine.hasGroupForArea("tap_head"));
        assertFalse(engine.hasGroupForArea("unknown_area"));
    }

    @Test
    void buildMotionCommand_forTapHead_returnsValidCommandEnvelope() {
        String json = engine.buildMotionCommand("tap_head");
        assertNotNull(json);

        Envelope envelope = Protocol.deserialize(json)
            .orElseThrow(() -> new AssertionError("Expected valid envelope JSON"));

        assertEquals("command", envelope.type());
        assertEquals("play_motion_ext", envelope.action());

        JsonObject payload = envelope.payload();
        assertTrue(payload.has("motion_path"));
        assertTrue(payload.has("priority"));
        assertTrue(payload.has("fade_in"));
        assertTrue(payload.has("fade_out"));
        assertEquals(2, payload.get("priority").getAsInt());

        String motionPath = payload.get("motion_path").getAsString();
        assertTrue(
            motionPath.endsWith("33.motion3.json") || motionPath.endsWith("34.motion3.json"),
            "Unexpected motion path: " + motionPath
        );
    }

    @Test
    void buildMotionCommand_forUnknownArea_returnsNull() {
        assertNull(engine.buildMotionCommand("unknown_area"));
    }

    @Test
    void buildMotionCommand_forAudioOnlyArea_returnsNull() {
        assertNull(engine.buildMotionCommand("audio_only"));
    }

    @Test
    void buildAudioOnlyCommand_forAudioOnlyArea_returnsPlayAudio() {
        String json = engine.buildAudioOnlyCommand("audio_only");
        assertNotNull(json);

        Envelope envelope = Protocol.deserialize(json)
            .orElseThrow(() -> new AssertionError("Expected valid envelope JSON"));

        assertEquals("command", envelope.type());
        assertEquals("play_audio", envelope.action());

        JsonObject payload = envelope.payload();
        assertNotNull(payload.get("audio_path"));
        assertTrue(payload.get("audio_path").getAsString().endsWith("1.ogg"),
            "Unexpected audio path: " + payload.get("audio_path"));
    }

    @Test
    void buildAudioOnlyCommand_forMotionGroup_returnsNull() {
        assertNull(engine.buildAudioOnlyCommand("tap_head"));
    }

    @Test
    void buildMotionCommand_convertsFadeMillisecondsToSeconds() {
        for (int i = 0; i < 20; i++) {
            String json = engine.buildMotionCommand("tap_head");
            assertNotNull(json);

            Envelope envelope = Protocol.deserialize(json)
                .orElseThrow(() -> new AssertionError("Expected valid envelope JSON"));

            float fadeIn = envelope.payload().get("fade_in").getAsFloat();
            float fadeOut = envelope.payload().get("fade_out").getAsFloat();

            boolean validFadeIn = Float.compare(fadeIn, 1.0f) == 0 || Float.compare(fadeIn, 0.5f) == 0;
            boolean validFadeOut = Float.compare(fadeOut, 1.0f) == 0 || Float.compare(fadeOut, 0.5f) == 0;

            assertTrue(validFadeIn, "fade_in should be seconds (0.5 or 1.0), got: " + fadeIn);
            assertTrue(validFadeOut, "fade_out should be seconds (0.5 or 1.0), got: " + fadeOut);
        }
    }
}

File: controller_qt/tests/MountedBehaviorEngineTest.cpp
// MountedBehaviorEngineTest — Phase 5 Wave 8 todo 20 TDD for the C++ port of
// Java's MountedBehaviorEngine (246 LOC). Five behaviors locked:
//
//   1. testBuildBehaviorCommand — synthetic VoicePackInfo with one head-tap
//      action (motion + audio) → buildBehaviorCommand returns a play_motion_ext
//      Envelope whose payload has absolute motion_path + audio_path +
//      subtitle_duration > 0 (from the OGG heuristic).
//   2. testHasGroupForAreaFalse — unknown area → false; case-insensitive
//      lookup (HEAD == head).
//   3. testNullVoicePack — default-constructed engine → hasGroupForArea always
//      false, buildBehaviorCommand returns nullopt (the "engine disabled"
//      sentinel that InstanceSession's hit handler relies on for the
//      InteractionHandler fallback).
//   4. testAudioOnly — action with empty motion + non-empty audio →
//      buildAudioOnlyCommand returns a play_audio Envelope.
//   5. testIsIdleMotionPath — an "idle" group action's resolved motion path is
//      recognized as idle; a non-matching path returns false.
//
// The estimateOggDurationMs heuristic has its own test file
// (OggDurationHeuristicTest.cpp) — it's a self-contained static method that
// parses the OGG binary format, distinct from the engine API cluster.

#include "core/MountedBehaviorEngine.hpp"

#include <QByteArray>
#include <QDir>
#include <QFile>
#include <QObject>
#include <QString>
#include <QTemporaryDir>
#include <QTest>

#include "network/Protocol.hpp"

namespace {

void appendLeU32(QByteArray& out, quint32 value)
{
    out.append(static_cast<char>(value & 0xFF));
    out.append(static_cast<char>((value >> 8) & 0xFF));
    out.append(static_cast<char>((value >> 16) & 0xFF));
    out.append(static_cast<char>((value >> 24) & 0xFF));
}

void appendLeU64(QByteArray& out, quint64 value)
{
    for (int i = 0; i < 8; ++i) {
        out.append(static_cast<char>((value >> (8 * i)) & 0xFF));
    }
}

// Minimal blob satisfying the OGG heuristic's two checkpoints: \x01vorbis
// header with sample_rate at marker+12, and a final OggS page with granule
// at +6. Only used here to populate the audio file in buildHeadPack — the
// heuristic ITSELF is tested in OggDurationHeuristicTest.cpp.
QByteArray synthesizeOggLikeBlob(quint32 sampleRate, quint64 granulePos)
{
    QByteArray out;
    out.append("OggS");
    out.append(8, '\0');
    out.append(4, '\0');
    out.append(4, '\0');
    out.append(4, '\0');
    out.append(static_cast<char>(1));
    out.append(static_cast<char>(30));
    out.append(static_cast<char>(0x01));
    out.append("vorbis");
    appendLeU32(out, 0);
    out.append(static_cast<char>(1));
    appendLeU32(out, sampleRate);
    out.append(12, '\0');
    out.append(64, '\0');
    out.append("OggS");
    out.append(static_cast<char>(0));
    out.append(static_cast<char>(0x04));
    appendLeU64(out, granulePos);
    out.append(4, '\0');
    out.append(4, '\0');
    out.append(4, '\0');
    out.append(static_cast<char>(0));
    return out;
}

// Build a VoicePackInfo with one group "head" containing a single action with
// motion + audio + lipSync. The files are written into tempDir.
core::VoicePackInfo buildHeadPack(const QTemporaryDir& tempDir)
{
    core::VoicePackInfo pack;
    pack.dirName = QStringLiteral("test-pack");
    pack.displayName = QStringLiteral("Test Pack");
    pack.code = QStringLiteral("test_pack");
    pack.basePath = tempDir.path();

    {
        QFile f(QDir(tempDir.path()).absoluteFilePath(QStringLiteral("a.ogg")));
        if (f.open(QIODevice::WriteOnly)) {
            f.write(synthesizeOggLikeBlob(44100, 44100));
        }
    }
    {
        QFile f(QDir(tempDir.path()).absoluteFilePath(QStringLiteral("m.json3")));
        if (f.open(QIODevice::WriteOnly)) {
            f.write("{\"Version\":3,\"Meta\":{\"Name\":\"m\"}}");
        }
    }
    {
        QFile f(QDir(tempDir.path()).absoluteFilePath(QStringLiteral("l.json")));
        if (f.open(QIODevice::WriteOnly)) {
            f.write("{}");
        }
    }

    core::VoicePackGroup group;
    group.code = QStringLiteral("head");
    group.name = QStringLiteral("Head");
    group.priority = 3;

    core::VoicePackAction action;
    action.id = 1;
    action.motionPath = QStringLiteral("m.json3");
    action.audioPath = QStringLiteral("a.ogg");
    action.lipSyncPath = QStringLiteral("l.json");
    action.doc = QStringLiteral("Ouch!");
    action.fadeInMs = 200;
    action.fadeOutMs = 300;
    group.actions.append(action);

    pack.groups.insert(QStringLiteral("head"), group);
    return pack;
}

} // namespace

class MountedBehaviorEngineTest : public QObject {
    Q_OBJECT

private slots:

    void testBuildBehaviorCommand()
    {
        QTemporaryDir tempDir;
        QVERIFY(tempDir.isValid());
        const auto pack = buildHeadPack(tempDir);
        const core::MountedBehaviorEngine engine(&pack);

        QVERIFY(engine.hasGroupForArea(QStringLiteral("head")));
        QVERIFY(engine.hasGroupForArea(QStringLiteral("HEAD")));

        const auto result = engine.buildBehaviorCommand(QStringLiteral("head"));
        QVERIFY(result.has_value());

        QCOMPARE(result->command.type, QStringLiteral("command"));
        QCOMPARE(result->command.action, QStringLiteral("play_motion_ext"));

        const QString motionPath = result->command.payload.value(
            QStringLiteral("motion_path")).toString();
        QVERIFY(!motionPath.isEmpty());
        QVERIFY(motionPath.endsWith(QStringLiteral("m.json3")));
        QVERIFY(QDir::isAbsolutePath(motionPath));

        const QString audioPath = result->command.payload.value(
            QStringLiteral("audio_path")).toString();
        QVERIFY(audioPath.endsWith(QStringLiteral("a.ogg")));
        QVERIFY(QDir::isAbsolutePath(audioPath));

        const QString lipSyncPath = result->command.payload.value(
            QStringLiteral("lip_sync_path")).toString();
        QVERIFY(lipSyncPath.endsWith(QStringLiteral("l.json")));

        QCOMPARE(result->command.payload.value(
            QStringLiteral("priority")).toInt(), 3);
        QCOMPARE(result->command.payload.value(
            QStringLiteral("fade_in")).toDouble(), 0.2);
        QCOMPARE(result->command.payload.value(
            QStringLiteral("fade_out")).toDouble(), 0.3);

        QCOMPARE(result->command.payload.value(
            QStringLiteral("subtitle_text")).toString(), QStringLiteral("Ouch!"));
        QCOMPARE(result->subtitleText, QStringLiteral("Ouch!"));

        QVERIFY(result->command.payload.contains(QStringLiteral("subtitle_duration")));
        QCOMPARE(result->audioDurationMs, 1000);
    }

    void testHasGroupForAreaFalse()
    {
        QTemporaryDir tempDir;
        QVERIFY(tempDir.isValid());
        const auto pack = buildHeadPack(tempDir);
        const core::MountedBehaviorEngine engine(&pack);

        QVERIFY(!engine.hasGroupForArea(QStringLiteral("body")));
        QVERIFY(!engine.hasGroupForArea(QStringLiteral("torso")));
        QVERIFY(!engine.hasGroupForArea(QString()));
        QVERIFY(!engine.buildBehaviorCommand(QStringLiteral("body")).has_value());
    }

    void testNullVoicePack()
    {
        const core::MountedBehaviorEngine engine;

        QVERIFY(!engine.hasGroupForArea(QStringLiteral("head")));
        QVERIFY(!engine.hasGroupForArea(QStringLiteral("body")));
        QVERIFY(!engine.buildBehaviorCommand(QStringLiteral("head")).has_value());
        QVERIFY(!engine.buildAudioOnlyCommand(QStringLiteral("head")).has_value());
        QVERIFY(!engine.isIdleMotionPath(QStringLiteral("C:/anything.json3")));
        QVERIFY(engine.groupNames().isEmpty());
    }

    void testAudioOnly()
    {
        QTemporaryDir tempDir;
        QVERIFY(tempDir.isValid());

        {
            QFile f(QDir(tempDir.path()).absoluteFilePath(QStringLiteral("voice.ogg")));
            QVERIFY(f.open(QIODevice::WriteOnly));
            f.write(synthesizeOggLikeBlob(22050, 22050));
        }

        core::VoicePackInfo pack;
        pack.dirName = QStringLiteral("audio-pack");
        pack.displayName = QStringLiteral("Audio");
        pack.code = QStringLiteral("audio_pack");
        pack.basePath = tempDir.path();

        core::VoicePackGroup group;
        group.code = QStringLiteral("body");
        group.name = QStringLiteral("Body");

        core::VoicePackAction motionAction;
        motionAction.motionPath = QStringLiteral("m.json3");
        group.actions.append(motionAction);

        core::VoicePackAction audioOnlyAction;
        audioOnlyAction.audioPath = QStringLiteral("voice.ogg");
        audioOnlyAction.doc = QStringLiteral("Hi there");
        group.actions.append(audioOnlyAction);

        pack.groups.insert(QStringLiteral("body"), group);

        const core::MountedBehaviorEngine engine(&pack);

        const auto motionResult = engine.buildBehaviorCommand(
            QStringLiteral("body"));
        QVERIFY(motionResult.has_value());
        QCOMPARE(motionResult->command.action, QStringLiteral("play_motion_ext"));

        const auto audioResult = engine.buildAudioOnlyCommand(
            QStringLiteral("body"));
        QVERIFY(audioResult.has_value());
        QCOMPARE(audioResult->action, QStringLiteral("play_audio"));

        const QString audioPath = audioResult->payload.value(
            QStringLiteral("audio_path")).toString();
        QVERIFY(audioPath.endsWith(QStringLiteral("voice.ogg")));
        QVERIFY(QDir::isAbsolutePath(audioPath));
        QCOMPARE(audioResult->payload.value(
            QStringLiteral("volume")).toDouble(), 1.0);
    }

    void testIsIdleMotionPath()
    {
        QTemporaryDir tempDir;
        QVERIFY(tempDir.isValid());

        core::VoicePackInfo pack;
        pack.dirName = QStringLiteral("idle-pack");
        pack.displayName = QStringLiteral("Idle");
        pack.code = QStringLiteral("idle_pack");
        pack.basePath = tempDir.path();

        core::VoicePackGroup idleGroup;
        idleGroup.code = QStringLiteral("idle");
        core::VoicePackAction a;
        a.motionPath = QStringLiteral("idle/m01.json3");
        idleGroup.actions.append(a);
        pack.groups.insert(QStringLiteral("idle"), idleGroup);

        const core::MountedBehaviorEngine engine(&pack);

        const QString resolved = QDir(tempDir.path()).absoluteFilePath(
            QStringLiteral("idle/m01.json3"));
        QVERIFY(engine.isIdleMotionPath(resolved));
        QVERIFY(!engine.isIdleMotionPath(
            QStringLiteral("C:/some/other/motion.json3")));
        QVERIFY(!engine.isIdleMotionPath(QString()));
    }
};

QTEST_APPLESS_MAIN(MountedBehaviorEngineTest)
#include "MountedBehaviorEngineTest.moc"
